pkg/common/log/logger.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package log
+
+import (
+	"sync"
+
+	"github.com/hyperledger/aries-framework-go/pkg/internal/common/logging/metadata"
+)
+
+const (
+	//loggerNotInitializedMsg is used when a logger is not initialized before logging
+	loggerNotInitializedMsg = "Default logger initialized (please call log.Initialize() if you wish to use a custom logger)"
+	loggerModule            = "aries-framework/common"
+)
+
+//Log is an implementation of Logger interface.
+//It encapsulates default or custom logger to provide module and level based logging.
+type Log struct {
+	instance Logger
+	module   string
+	once     sync.Once
+}
+
+// NewLogger creates and returns a Logger object based on the given module name.
+// note: the underlying logger instance is lazy initialized on first use.
+// To use your own logger implementation provide logger provider in 'Initialize()' before logging any line.
+// If 'Initialize()' is not called before logging any line then default logging implementation will be used.
+func NewLogger(module string) Logger {
```

Comment:
Interesting. It is log.NewLogger since it returns a Logger.

But under the covers it is actually log.NewLog (which could be returning *Log).

i.e., should we have the extra abstraction overhead?

Reply:
CC: @bstasyszyn

Reply:
Since this is returning implementation of 'Logger' interface, I named in NewLogger (also maintaining naming conventions used in other kind of loggers)

Reply:
actually i should have said log.New rather than log.NewLog (like the stdlib)

Reply:
Fixed it, changed return to *Log